File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/CSharp/Silverlight/2052/BusinessApplication/Models/RegistrationData.partial.cs
namespace $safeprojectname$.Web
{
    using System;
    using System.ComponentModel.DataAnnotations;
    using System.ServiceModel.DomainServices.Client;
    using System.ServiceModel.DomainServices.Client.ApplicationServices;
    using $safeprojectname$.Web.Resources;

    /// <summary>
    /// 用于提供客户端自定义验证和对 <see cref="RegistrationData"/> 的数据绑定的扩展。
    /// </summary>
    public partial class RegistrationData
    {
        private OperationBase currentOperation;

        /// <summary>
        /// 获取或设置返回密码的函数。
        /// </summary>
        internal Func<string> PasswordAccessor { get; set; }

        /// <summary>
        /// 获取和设置密码。
        /// </summary>
        [Required(ErrorMessageResourceName = "ValidationErrorRequiredField", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        [Display(Order = 3, Name = "PasswordLabel", Description = "PasswordDescription", ResourceType = typeof(RegistrationDataResources))]
        [RegularExpression("^.*[^a-zA-Z0-9].*$", ErrorMessageResourceName = "ValidationErrorBadPasswordStrength", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        [StringLength(50, MinimumLength = 7, ErrorMessageResourceName = "ValidationErrorBadPasswordLength", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        public string Password
        {
            get
            {
                return (this.PasswordAccessor == null) ? string.Empty : this.PasswordAccessor();
            }

            set
            {
                this.ValidateProperty("Password", value);
                this.CheckPasswordConfirmation();

                // 请不要将密码存储在私有字段中，因为不应以纯文本形式在内存中存储密码。
                // 而应将提供的 PasswordAccessor 用作该值的后备存储。

                this.RaisePropertyChanged("Password");
            }
        }

        /// <summary>
        /// 获取或设置返回密码确认的函数。
        /// </summary>
        internal Func<string> PasswordConfirmationAccessor { get; set; }

        /// <summary>
        /// 获取和设置密码确认字符串。
        /// </summary>
        [Required(ErrorMessageResourceName = "ValidationErrorRequiredField", ErrorMessageResourceType = typeof(ValidationErrorResources))]
        [Display(Order = 4, Name = "PasswordConfirmationLabel", ResourceType = typeof(RegistrationDataResources))]
        public string PasswordConfirmation
        {
            get
            {
                return (this.PasswordConfirmationAccessor == null) ? string.Empty : this.PasswordConfirmationAccessor();
            }

            set
            {
                this.ValidateProperty("PasswordConfirmation", value);
                this.CheckPasswordConfirmation();

                // 请不要将密码存储在私有字段中，因为不应以纯文本形式在内存中存储密码。
                // 而应将提供的 PasswordAccessor 用作该值的后备存储。

                this.RaisePropertyChanged("PasswordConfirmation");
            }
        }

        /// <summary>
        /// 获取或设置当前注册或登录操作。
        /// </summary>
        internal OperationBase CurrentOperation
        {
            get
            {
                return this.currentOperation;
            }
            set
            {
                if (this.currentOperation != value)
                {
                    if (this.currentOperation != null)
                    {
                        this.currentOperation.Completed -= (s, e) => this.CurrentOperationChanged();
                    }

                    this.currentOperation = value;

                    if (this.currentOperation != null)
                    {
                        this.currentOperation.Completed += (s, e) => this.CurrentOperationChanged();
                    }

                    this.CurrentOperationChanged();
                }
            }
        }

        /// <summary>
        /// 获取一个值，该值指示用户当前是否正在注册或登录。
        /// </summary>
        [Display(AutoGenerateField = false)]
        public bool IsRegistering
        {
            get
            {
                return this.CurrentOperation != null && !this.CurrentOperation.IsComplete;
            }
        }

        /// <summary>
        /// 在当前操作发生更改时要使用的帮助程序方法。
        /// 用于引发合适的属性更改通知。
        /// </summary>
        private void CurrentOperationChanged()
        {
            this.RaisePropertyChanged("IsRegistering");
        }

        /// <summary>
        /// 检查以确保密码与确认密码匹配。
        /// 如果两者不匹配，则添加验证错误。
        /// </summary>
        private void CheckPasswordConfirmation()
        {
            // 如果未输入两个密码中的任何一个，则不测试这两个字段的相等性。
            // Required 特性确保为这两个字段都输入了值。
            if (string.IsNullOrWhiteSpace(this.Password)
                || string.IsNullOrWhiteSpace(this.PasswordConfirmation))
            {
                return;
            }

            // 如果两个值不同，则添加一个验证错误，并同时指定这两个成员。
            if (this.Password != this.PasswordConfirmation)
            {
                this.ValidationErrors.Add(new ValidationResult(ValidationErrorResources.ValidationErrorPasswordConfirmationMismatch, new string[] { "PasswordConfirmation", "Password" }));
            }
        }

        /// <summary>
        /// 在输入 UserName 值之后执行逻辑。
        /// </summary>
        /// <param name="userName">输入的用户名值。</param>
        /// <remarks>
        /// 允许窗体指示值在何时已输入完整。
        /// 使用 OnUserNameChanged 方法可能导致在用户在窗体中输入完整值之前过早进行调用。
        /// </remarks>
        internal void UserNameEntered(string userName)
        {
            // 在没有指定友好名称时为新实体自动填充 FriendlyName 以匹配 UserName
            if (string.IsNullOrWhiteSpace(this.FriendlyName))
            {
                this.FriendlyName = userName;
            }
        }

        /// <summary>
        /// 创建使用此实体的数据(IsPersistent 默认为 false)初始化的新 <see cref="LoginParameters"/>。
        /// </summary>
        public LoginParameters ToLoginParameters()
        {
            return new LoginParameters(this.UserName, this.Password, false, null);
        }
    }
}

File: src/ChpokkWeb/SystemFiles/Templates/ProjectTemplates/VisualBasic/Silverlight/1042/BusinessApplication/Models/RegistrationData.partial.vb
Imports System
Imports System.ComponentModel
Imports System.ComponentModel.DataAnnotations
Imports System.ServiceModel.DomainServices.Client
Imports System.ServiceModel.DomainServices.Client.ApplicationServices

Namespace Web

    Partial Public Class RegistrationData
        Private _CurrentOperation As OperationBase

        ''' <summary>
        ''' 암호를 반환하는 함수를 가져오거나 설정합니다.
        ''' </summary>
        Friend Property PasswordAccessor() As Func(Of String)

        ''' <summary>
        ''' 암호를 가져오거나 설정합니다.
        ''' </summary>
        <Required(ErrorMessageResourceName:="ValidationErrorRequiredField", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        <Display(Order:=3, Name:="PasswordLabel", Description:="PasswordDescription", ResourceType:=GetType(RegistrationDataResources))> _
        <RegularExpression("^.*[^a-zA-Z0-9].*$", ErrorMessageResourceName:="ValidationErrorBadPasswordStrength", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        <StringLength(50, MinimumLength:=7, ErrorMessageResourceName:="ValidationErrorBadPasswordLength", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        Public Property Password() As String
            Get
                Return If((Me.PasswordAccessor Is Nothing), String.Empty, Me.PasswordAccessor.Invoke())
            End Get

            Set(ByVal value As String)
                Me.ValidateProperty("Password", value)
                Me.CheckPasswordConfirmation()

                ' 암호를 메모리에 일반 텍스트로 저장해서는 안 되므로 암호를 전용 필드에 저장하지 마십시오.
                ' 대신, 제공된 PasswordAccessor가 값에 대한 백업 저장소 역할을 합니다.

                Me.RaisePropertyChanged("Password")
            End Set
        End Property

        ''' <summary>
        ''' 암호 확인을 반환하는 함수를 가져오거나 설정합니다.
        ''' </summary>
        Friend Property PasswordConfirmationAccessor() As Func(Of String)

        ''' <summary>
        ''' 암호 확인 문자열을 가져오거나 설정합니다.
        ''' </summary>
        <Required(ErrorMessageResourceName:="ValidationErrorRequiredField", ErrorMessageResourceType:=GetType(ValidationErrorResources))> _
        <Display(Order:=4, Name:="PasswordConfirmationLabel", ResourceType:=GetType(RegistrationDataResources))> _
        Public Property PasswordConfirmation() As String
            Get
                Return If((Me.PasswordConfirmationAccessor Is Nothing), String.Empty, Me.PasswordConfirmationAccessor.Invoke())
            End Get

            Set(ByVal value As String)
                Me.ValidateProperty("PasswordConfirmation", value)
                Me.CheckPasswordConfirmation()

                ' 암호를 메모리에 일반 텍스트로 저장해서는 안 되므로 암호를 전용 필드에 저장하지 마십시오.
                ' 대신, 제공된 PasswordAccessor가 값에 대한 백업 저장소 역할을 합니다.

                Me.RaisePropertyChanged("PasswordConfirmation")
            End Set
        End Property

        ''' <summary>
        ''' 현재 등록 또는 로그인 작업을 가져오거나 설정합니다.
        ''' </summary>
        Friend Property CurrentOperation() As OperationBase
            Get
                Return _CurrentOperation
            End Get
            Set(ByVal value As OperationBase)
                If Not Object.Equals(_CurrentOperation, value) Then
                    If Not IsNothing(_CurrentOperation) Then
                        RemoveHandler _CurrentOperation.Completed, AddressOf Me.CurrentOperationChanged
                    End If

                    _CurrentOperation = value

                    If Not IsNothing(_CurrentOperation) Then
                        AddHandler _CurrentOperation.Completed, AddressOf Me.CurrentOperationChanged
                    End If

                    Me.CurrentOperationChanged(Me, EventArgs.Empty)
                End If
            End Set
        End Property

        ''' <summary>
        ''' 사용자가 현재 등록되어 있거나 로그인해 있는지 여부를 나타내는 값을 가져옵니다.
        ''' </summary>
        <Display(AutoGenerateField:=False)> _
        Public ReadOnly Property IsRegistering() As Boolean
            Get
                Return (Not IsNothing(Me.CurrentOperation)) AndAlso (Not Me.CurrentOperation.IsComplete)
            End Get
        End Property

        ''' <summary>
        ''' 현재 작업이 변경된 경우에 대한 도우미 메서드입니다.
        ''' 적절한 속성 변경 알림을 표시하는 데 사용됩니다.
        ''' </summary>
        ''' <param name="sender">이벤트 전송자입니다.</param>
        ''' <param name="e">event 인수입니다.</param>
        Private Sub CurrentOperationChanged(ByVal sender As Object, ByVal e As EventArgs)
            Me.RaisePropertyChanged("IsRegistering")
        End Sub

        ''' <summary>
        ''' 암호와 암호 확인이 일치하는지 확인합니다.
        ''' 일치하지 않는 경우 유효성 검사 오류가 추가됩니다.
        ''' </summary>
        Private Sub CheckPasswordConfirmation()
            ' 암호 또는 암호 확인을 입력하지 않은 경우 두 필드가 일치하는지 테스트하지 않습니다.
            ' Required 특성은 두 필드 모두에 값이 입력되었는지 확인합니다.
            If String.IsNullOrWhiteSpace(Me.Password) OrElse String.IsNullOrWhiteSpace(Me.PasswordConfirmation) Then
                Exit Sub
            End If

            ' 값이 다른 경우 두 멤버를 지정하여 유효성 검사 오류를 추가합니다.
            If Me.Password <> Me.PasswordConfirmation Then
                Me.ValidationErrors.Add(New ValidationResult(ValidationErrorResources.ValidationErrorPasswordConfirmationMismatch, New String() {"PasswordConfirmation", "Password"}))
            End If
        End Sub

        ''' <summary>
        ''' UserName 값이 입력된 이후에 논리를 수행합니다.
        ''' </summary>
        ''' <param name="userName">입력된 사용자 이름입니다.</param>
        ''' <remarks>
        ''' 폼을 사용하여 값이 입력된 시간을 나타낼 수 있습니다.
        ''' OnUserNameChanged 메서드를 사용하면 사용자가 폼에 값을 입력하기 전에 호출이 발생할 수 있습니다.
        ''' </remarks>
        Friend Sub UserNameEntered(ByVal userName As String)
            ' 표시 이름을 지정하지 않은 경우 새 엔터티에 대한 UserName과 일치하도록 FriendlyName을 자동으로 채웁니다.
            If String.IsNullOrWhiteSpace(Me.FriendlyName) Then
                Me.FriendlyName = userName
            End If
        End Sub

        ''' <summary>
        ''' 이 엔터티의 데이터로 초기화되는 새 <see cref="LoginParameters"/>를 만듭니다. IsPersistent의 기본값은 false입니다.
        ''' </summary>
        Public Function ToLoginParameters() As LoginParameters
            Return New LoginParameters(Me.UserName, Me.Password, False, Nothing)
        End Function
    End Class
End Namespace